### PR TITLE
Re authorize payment

### PR DESCRIPTION
**Pull request by jdungan**
Includes:
1) clean up of web hook receipt
2) adds verification of events with stripe to fix CSRF exempt worries
3) create command for reauthorizing about to expire charges.

### CODE REVIEW

File: auctions/models.py [hunk 1]
```diff
@@ -573,13 +578,11 @@ def __unicode__(self):
...
 @receiver(post_save, sender=Bid)
 @receiver(post_save, sender=Claim)
 @receiver(post_save, sender=Vote)
 def update_datetimes_for_model_save(sender, instance, created, **kwargs):
     if created:
-        sender.objects.filter(id=instance.id).update(created=datetime.now())
+        sender.objects.filter(id=instance.id).update(created=timezone.now())
     else:
-        sender.objects.filter(id=instance.id).update(modified=datetime.now())
+        sender.objects.filter(id=instance.id).update(modified=timezone.now())
```
Comment (groovecoder) [inline auctions/models.py]: There's another use of `datetime.now()` in `notify_matching_askers`. I'll file an issue to replace all (or as many as possible) uses of `datetime.now()` with `timezone.now()`.

Comment (groovecoder) [inline auctions/models.py]: Filed https://github.com/codesy/codesy/issues/397

File: payments/admin.py [hunk 2]
```diff
@@ -1,6 +1,29 @@
...
+
 from django.contrib import admin
-from payments.models import StripeAccount, StripeEvent
+from .models import StripeAccount, StripeEvent
+
+
+class StripeAccountAdmin(admin.ModelAdmin):
+
+    readonly_fields = (
+        'user', 'account_id', 'secret_key', 'public_key',
```
Comment (groovecoder) [inline payments/admin.py]: Do we need to see the `secret_key` values here? Or can we keep it secret such that it requires connecting to the database to see it?

File: payments/admin.py [hunk 3]
```diff
@@ -1,6 +1,29 @@
...
+
+class StripeEventAdmin(admin.ModelAdmin):
+    list_display = (
+        'event_id', 'type', 'created', 'verified', 'processed', 'user_id')
+    search_fields = ['message_text']
+
+    readonly_fields = (
+        'type', 'created', 'event_id', 'processed', 'message_text',)
 
+    def message_text(self, instance):
```
Comment (groovecoder) [inline payments/admin.py]: And if we're worried about displaying secrets in the django admin interface, maybe we want to strip any secrets out of message before we display it?

File: payments/management/commands/reauthorize.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+
+class Command(BaseCommand):
+
+    def add_arguments(self, parser):
+        parser.add_argument('day_limit', nargs='+', type=int)
+
+    def handle(self, *args, **options):
+        day_limit = options['day_limit'][0]
+
```
Comment (groovecoder) [inline payments/management/commands/reauthorize.py]: let's add a `#TODO` above this to move the expiration date calculation into the `filter` to only return offers that will expire.

File: payments/models.py [hunk 5]
```diff
@@ -27,8 +28,16 @@ class StripeAccount(models.Model):
     verification = models.TextField(default='', blank=True)
 
     def identity_verified(self):
+        if not self.verification:
+            return True
+        else:
+            verification = json.loads(self.verification)
+        return not verification['due_by']
```
Comment (groovecoder) [inline payments/models.py]: Let's indent this to move it into the `else` block.

File: payments/models.py [hunk 6]
```diff
@@ -57,52 +66,102 @@ def external_account(self, bank_account):
 
 class StripeEvent(models.Model):
     event_id = models.CharField(primary_key=True, max_length=100, blank=True)
+    user_id = models.CharField(max_length=100, blank=True)
     type = models.CharField(max_length=100, blank=True)
     message_text = models.TextField()
+    verified = models.BooleanField(default=False)
     processed = models.BooleanField(default=False)
+    created = models.DateTimeField(null=True, blank=True)
```
Comment (groovecoder) [inline payments/models.py]: Can we change this (and other `DateTimeField`s to use `auto_now_add=True`). See https://docs.djangoproject.com/en/1.10/ref/models/fields/#django.db.models.DateField.auto_now

File: payments/models.py [hunk 7]
```diff
@@ -57,52 +66,102 @@ def external_account(self, bank_account):
...
+            retrieved_event = stripe.Event.retrieve(
+                id=self.event_id, stripe_account=self.user_id)
         except KeyError:
+            retrieved_event = stripe.Event.retrieve(id=self.event_id)
+
+        if retrieved_event:
+            self.verified = True
+            self.type = retrieved_event['type']
+            self.message_text = json.dumps(retrieved_event)
+        super(StripeEvent, self).save(*args, **kwargs)
```
Comment (groovecoder) [inline payments/models.py]: This `super.save()` should move into the `if` block, so we only save if there really was an event from Stripe. And we should add an `else` block which bails out and/or logs a message like, "Could not retrieve Stripe Event ID: <id>"

File: payments/models.py [hunk 8]
```diff
@@ -57,52 +66,102 @@ def external_account(self, bank_account):
...
+
+        # attempt to process a webhook for the event type
+        try:
+            webhooks[self.type](event=self).process()
+            self.processed = True
```
Comment (groovecoder) [inline payments/models.py]: Should this `self.processed = True` move to the end of the `process()` method itself?

File: payments/models.py [hunk 9]
```diff
@@ -57,52 +66,102 @@ def external_account(self, bank_account):
...
+
+    def process():
+        raise NotImplementedError
+
+
+class AccountUpdatedProcessor(WebHookProcessor):
+    def process(self):
         try:
-            account = StripeAccount.objects.get(account_id=account_id)
+            account = StripeAccount.objects.get(account_id=self.account_id)
```
Comment (groovecoder) [inline payments/models.py]: I think this is supposed to be `self.object['account_id']` or `self.object.account_id`?

File: payments/models.py [hunk 10]
```diff
@@ -57,52 +66,102 @@ def external_account(self, bank_account):
...
 
-class BalanceAvailableProcessor(object):
-    def process(self, event):
-        message = event.message()
-        account_id = message['user_id']
+class BalanceAvailableProcessor(WebHookProcessor):
+    def process(self):
         try:
-            account = StripeAccount.objects.get(account_id=account_id)
+            account = StripeAccount.objects.get(account_id=self.account_id)
```
Comment (groovecoder) [inline payments/models.py]: Same here.

File: payments/views.py [hunk 11]
```diff
@@ -53,24 +53,21 @@ def dispatch(self, *args, **kwargs):
...
+        return self.request.user.accepted_terms()
 
 
-class UserHasAcceptedTermsMixin(UserPassesTestMixin):
-    login_url = reverse_lazy('terms')
+class UserIdentityVerifiedMixin(UserHasAcceptedTermsMixin):
+    login_url = reverse_lazy('identity')
     redirect_field_name = None
 
     def test_func(self):
```
Comment (groovecoder) [inline payments/views.py]: Need to call `super().test_func()` to make `UserIdentityVerifiedMixin` actually stack ON TOP of `UserHasAcceptedTermsMixin`

**Comment by groovecoder** [comment]
Looking good!! Few comments.

File: payments/admin.py (hunk 2 above)
Comment (jdungan) [inline payments/admin.py]: Setting the model to editable=False and removing it from the readyonly_fields list hides the secret_key from admin.

File: payments/admin.py (hunk 3 above)
Comment (jdungan) [inline payments/admin.py]: The manage user secret key is only sent when the account is created.  Pretty sure it is never sent with an event message.  From https://stripe.com/docs/connect/managed-accounts:

"You’ll want to store the account ID and, if you intend to use them for authentication, the secret key and publishable key. The keys will not be retrievable subsequently."

File: payments/models.py (hunk 6 above)
Comment (jdungan) [inline payments/models.py]: Added issue: #398

File: payments/models.py (hunk 8 above)
Comment (jdungan) [inline payments/models.py]: It wouldn't be pretty.

File: payments/models.py (hunk 9 above)
Comment (jdungan) [inline payments/models.py]: account = StripeAccount.objects.get(account_id=self.object['id'])

File: payments/views.py (hunk 11 above)
Comment (jdungan) [inline payments/views.py]: Super does not work, in fact trying to stack mixins doesn't work.  By definition each mixins are setting the login_url property.  I decide to write one mixin that test both conditions.

File: payments/models.py (hunk 7 above)
Comment (jdungan) [inline payments/models.py]: 6efc96a

File: payments/models.py (hunk 5 above)
Comment (jdungan) [inline payments/models.py]: 6efc96a

File: codesy/base/models.py [hunk 12]
```diff
@@ -25,7 +25,11 @@ def get_gravatar_url(self):
                 github_account.uid)
 
     def accepted_terms(self):
-        return self.tos_acceptance_date and self.tos_acceptance_ip
+        return (
+            not (not self.tos_acceptance_date)
+            and
+            not (not self.tos_acceptance_ip)
+        )
```
Comment (groovecoder) [inline codesy/base/models.py]: Boolean complexity here makes me go cross-eyed.

File: payments/models.py [hunk 13]
```diff
@@ -20,15 +21,23 @@ def get_customer_token(user):
 class StripeAccount(models.Model):
     user = models.ForeignKey(settings.AUTH_USER_MODEL)
     account_id = models.CharField(max_length=100, blank=True)
-    secret_key = models.CharField(max_length=100, blank=True)
+    secret_key = models.CharField(max_length=100, blank=True, editable=False)
```
Comment (groovecoder) [inline payments/models.py]: I haz a happy!

File: payments/models.py [hunk 14]
```diff
@@ -57,52 +66,106 @@ def external_account(self, bank_account):
...
+            self.user_id = self.message_text['user_id']
+            retrieved_event = stripe.Event.retrieve(
+                id=self.event_id, stripe_account=self.user_id)
+        except KeyError:
+            retrieved_event = stripe.Event.retrieve(id=self.event_id)
+
+        if retrieved_event:
+            self.verified = True
+            self.type = retrieved_event['type']
+            # indent used to avoid c encoder bug
```
Comment (groovecoder) [inline payments/models.py]: Wow. Grab the bug url for that and we'll put money on it.

File: payments/templates/verify_identity.html [hunk 15]
```diff
@@ -8,49 +8,63 @@
              We'll try to keep this to a minimum and do not save this information in our database.
          </p>
         <fieldset class="fieldset">
-          <legend>.</legend>
         <form
             role="form"
             class="form-horizontal"
-            action="/payments/identity"
+t            action="/payments/identity"
```
Comment (groovecoder) [inline payments/templates/verify_identity.html]: ? stray `t` character snuck in here? `flake8` doesn't care?

File: payments/views.py [hunk 16]
```diff
@@ -53,24 +53,35 @@ def dispatch(self, *args, **kwargs):
...
+    def test_func(self):
+        return self.request.user.accepted_terms()
+
+
 class UserIdentityVerifiedMixin(UserPassesTestMixin):
     login_url = reverse_lazy('identity')
     redirect_field_name = None
 
-    def test_func(self):
+    def test_func(self, **kwargs):
```
Comment (groovecoder) [inline payments/views.py]: Do we need to have `**kwargs` in the method parameter signature here?

File: payments/views.py [hunk 17]
```diff
@@ -53,24 +53,35 @@ def dispatch(self, *args, **kwargs):
...
 
 
-class UserHasAcceptedTermsMixin(UserPassesTestMixin):
+class BankAccountTestsMixin(UserPassesTestMixin):
     login_url = reverse_lazy('terms')
     redirect_field_name = None
 
-    def test_func(self):
-        return self.request.user.accepted_terms()
+    def test_func(self, **kwargs):
```
Comment (groovecoder) [inline payments/views.py]: Do we need to capture these `**kwargs` method params?

**Comment by groovecoder** [comment]
Few more comments.

File: payments/models.py (hunk 14 above)
Comment (jdungan) [inline payments/models.py]: https://github.com/stripe/stripe-python/issues/220

**Pull request merged**